File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */


var characterReplacement = function(s, k) {
    const charCounter = {};
    let longestStrLength = -Infinity;


    let [left, right] = [0, 0]
    charCounter[s[left]] = 1;
    while(right < s.length) {
        const windowLength = (right - left) + 1;
        const maxCount = Math.max(...Object.values(charCounter))
        const maxChanges = windowLength - maxCount;


        if (maxChanges <= k) {
            longestStrLength = Math.max(longestStrLength, windowLength);
            right++;
            charCounter[s[right]] = (charCounter[s[right]] || 0) + 1;
        } else {
            charCounter[s[left]] -= 1;
            left++;
        }
    }


    return Math.max(longestStrLength || 0)
};

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        longestStr = float("-inf")
        charCounter = {}
        left, right = 0, 0
        charCounter[s[right]] = 1

        while right < len(s):
            windowLen = right - left + 1
            maxCount = max(charCounter.values())
            maxChanges = windowLen - maxCount

            if maxChanges <= k:
                longestStr = max(longestStr, windowLen)
                right += 1
                print("right", right)
                if right < len(s):
                    charCounter[s[right]] = charCounter.get(s[right], 0) + 1
            else:
                charCounter[s[left]] -= 1
                left += 1
        
        return max(longestStr, 0)

        